Public/Src/Cache/ContentStore/Distributed/Sessions/ReadOnlyDistributedContentSession.cs

```diff
@@ -360,17 +360,53 @@ async Task<BoolResult> TryCopyContentLocalAsync()
         }
 
         /// <inheritdoc />
-        protected override async Task<IEnumerable<Task<Indexed<PinResult>>>> PinCoreAsync(OperationContext operationContext, IReadOnlyList<ContentHash> contentHashes, UrgencyHint urgencyHint, Counter retryCounter, Counter fileCounter)
+        public Task<IEnumerable<Task<Indexed<PinResult>>>> PinAsync(OperationContext operationContext, IReadOnlyList<ContentHash> contentHashes, PinOperationConfiguration pinOperationConfiguration)
+        {
+            return PinHelperAsync(operationContext, contentHashes, UrgencyHint.Nominal, pinOperationConfiguration);
+        }
+
+        /// <inheritdoc />
+        protected override Task<IEnumerable<Task<Indexed<PinResult>>>> PinCoreAsync(OperationContext operationContext, IReadOnlyList<ContentHash> contentHashes, UrgencyHint urgencyHint, Counter retryCounter, Counter fileCounter)
+        {
+            return PinHelperAsync(operationContext, contentHashes, urgencyHint, PinOperationConfiguration.Default());
+        }
+
+        private async Task<IEnumerable<Task<Indexed<PinResult>>>> PinHelperAsync(OperationContext operationContext, IReadOnlyList<ContentHash> contentHashes, UrgencyHint urgencyHint, PinOperationConfiguration pinOperationConfiguration)
         {
             Contract.Requires(contentHashes != null);
 
-            return await Workflows.RunWithFallback(
-                contentHashes,
-                hashes => Inner.PinAsync(operationContext, hashes, operationContext.Token, urgencyHint),
-                hashes => _remotePinner(operationContext, hashes, operationContext.Token, urgencyHint),
-                result => result.Succeeded,
-                // Exclude the empty hash because it is a special case which is hard coded for place/openstream/pin.
-                async hits => await UpdateContentTrackerWithLocalHitsAsync(operationContext, hits.Where(x => !(Settings.EmptyFileHashShortcutEnabled && contentHashes[x.Index].IsEmptyHash())).Select(x => new ContentHashWithSizeAndLastAccessTime(contentHashes[x.Index], x.Item.ContentSize, x.Item.LastAccessTime)).ToList(), operationContext.Token, urgencyHint));
+            IEnumerable<Task<Indexed<PinResult>>> pinResults = null;
+
+            if (pinOperationConfiguration.ReturnGlobalExistenceFast)

```

Comment:
I think this should be done in the second callback provided to RunWithFallback.

Currently, you go to the remote first without trying to pin locally.